### ISSUE TITLE
Support webgpu 

### ISSUE THREAD

**Issue by happydpc** [StarArawn/harmony#44]
This project is very promising like rendy's architecture. If it will support wasm, then will be better. The wasm may limit the crossbeam usage, it may be a problem.

**Comment by StarArawn** [comment]
Currently our rendering architecture is very multi-threaded. We use legion which currently doesn't have paralllel wasm support:
https://github.com/TomGillen/legion#wasm

It might be possible to run harmony on a single thread, but I'm not sure how well that will perform. Always open to seeing PR's for wasm support though! 😄

**Comment by chemicstry** [comment]
I've worked on implemeting wasm support for amethyst engine and it is quite simple, because you can create a rayon-like wrapper for web workers (https://github.com/amethyst/web_worker). HOWEVER, there is on big limitation that you can't sleep (mutex lock) the main thread and it breaks a lot of existing code. For example, if you want to wait for legion scheduler to finish, you have to return main thread to the browser and then wake up from javascript event (web worker message for example) to continue execution. Or run everything in web workers, but that's another can of worms (rendering works, audio doesn't (yet)). I may take a look some day into what's needed to implement it.

**Comment by StarArawn** [comment]
Yeah, another challenge is the new asset system I wrote. Currently it's very threadpool heavy and uses a concurrent hashmap. I think you could feature gate it though and block on assets loading instead of loading them on the thread pool.

**Comment by StarArawn** [comment]
This is further blocked by our use of push constants now. These do appear to be something that should be supported via a web extension in the future though.
